# Measuring interfacial Dzyaloshinskii-Moriya interaction in ultra-thin magnetic films
## M. Kuepferling, A. Casiraghi, G. Soares, G. Durin
Istituto Nazionale di Ricerca Metrologica, Strada delle Cacce 91, 10135, Turin, Italy

## F. Garcia-Sanchez
Dep. of Applied Physics, University of Salamanca, Plaza de los Caidos s/n 37008, Salamanca, Spain

## L. Chen, C.H. Back
Technical University Munich, James-Frank-Str. 1, 85748 Garching, Germany

## C.H. Marrows
School of Physics and Astronomy, University of Leeds, Leeds LS2 9JT, United Kingdom

## S. Tacchi
CNR, Istituto Officina dei Materiali - Perugia, c/o Dipartimento di Fisica e Geologia\\Università Perugia, Via A. Pascoli, 06123 Perugia, Italy}

## G.Carlotti
Dipartimento di Fisica e Geologia, Università di Perugia, Via Pascoli, 06123 Perugia, Italy

The Dzyaloshinskii-Moriya interaction (DMI), being one of the origins of chiral magnetism, is currently attracting considerable attention in the research community focusing on applied magnetism and spintronics. For future applications, an accurate measurement of its strength is indispensable. Here we present a review of the state-of-the-art of measuring the coefficient of the Dzyaloshinskii-Moriya interaction, the DMI constant D, focusing on systems where the interaction arises from the interface between two materials (i.e. interfacial DMI). We give an overview of the experimental techniques as well as their theoretical background and models for the quantification of the DMI constant. The measurement techniques are divided into three categories: a) domain wall-based measurements, b) spin wave-based measurements and c) spin-orbit torque-based measurements. We analyze the advantages and disadvantages of each method and compare D values at different interfaces. The review aims to obtain a better understanding of the applicability of the different techniques to various stacks and of the origin of apparent disagreements among literature values.

To be published on Review of Modern Physics   
Pre-print available [on ArXiv](https://arxiv.org/abs/2009.11830)

## This notebook generates figs. 28 and 29 of the paper, and other statistics

### Author: G. Durin

#### Version: 0.9.1 Date: Oct 6, 2022

Version for the accepted paper on Rev. of Modern Physics. First version published on GitHub


### Notes

Running this notebook in JupyterLab or as old Notebook should give an error when a package is not installed.   
Simply "pip install package" and restart the kernel

The plots make use of bokeh, https://bokeh.org/

The annular plots are based on this example: https://docs.bokeh.org/en/latest/docs/gallery/burtin.html

*If you want to add your recent data on DMI in a thin film, please write to g.durin@inrim.it*

In [1]:
import io
from IPython.display import display, Latex, HTML
import numpy as np
import pandas as pd
import Analyse_DMI as admi

In [2]:
from bokeh.io import push_notebook, show, output_notebook, export_png
from bokeh.layouts import column, row, gridplot
from bokeh.plotting import figure, output_file, save, show
from bokeh.models import Label
from bokeh.resources import CDN
from bokeh.embed import file_html, components
from bokeh.util.warnings import BokehUserWarning 
import warnings 
warnings.simplefilter(action='ignore', category=BokehUserWarning)
output_notebook()

Loading BokehJS ...

# The DMI database

The DMI values have been extracted from the literature and reported in the included Excel file "DMI.xlsx".    
Each sheet contains a different method used to evaluate the DMI, as shown below. The sheetname in the Excel file is the shortname below.

In [3]:
available_methods = pd.read_excel("DMI_methods_description.xlsx", index_col=0)
available_methods

Method Fullname  Chapter
Method Shortname                                                            
DW_current                        Current-driven domain wall motion   II.B.1
DW_field_creep       Field-driven domain wall motion (creep regime)   II.B.2
DW_field_flow         Field-driven domain wall motion (flow regime)   II.B.2
Domain_pattern                    Equilibrium stripe domain pattern   II.B.3
Stripe_annihilation                    Magnetic stripe annihilation   II.B.4
Nucleation                                         Nucleation field   II.B.5
Stray_field                                Domain wall stray fields   II.B.6
BLS                                      Brillouin light scattering  III.C.1
TR_MOKE                       Time resolved magneto-optical imaging  III.C.2
PSWS                             Propagating spin wave spectroscopy  III.C.3
MASW                                      Magneto-acustic spin wave  III.C.3
Spin_orbit_torque                                 Spin Orbit Torque     IV.C

## This is an example of data available

In [4]:
# Filename of the DMI data (excel) 
filename_DMI = "DMI.xlsx"
data = pd.ExcelFile(filename_DMI, engine='openpyxl')
# See an example
data.parse("DW_field_creep",dtype={"composition": str}).fillna("").iloc[21:30]

fm  tfm composition bottom1 tbottom1 bottom  tbottom  top  ttop top1  \
21     Co  0.9                                  Pt      5.0   Au   5.0        
22     Co  0.9                                  Au      5.0   Pt   5.0        
23  CoFeB  0.8      206020                      Ta      5.0  MgO   2.0        
24  CoFeB  1.0      206020                       W      2.0  MgO   2.0        
25  CoFeB  1.0      206020                       W      3.0  MgO   2.0        
26  CoFeB  1.0      206020                     TaN      1.0  MgO   2.0        
27  CoFeB  1.0      206020                      Hf      1.0  MgO   2.0        
28  CoFeB  0.6      206020                       W      5.0  MgO   2.0        
29  CoFeB  0.6      404020                       W      5.0  MgO   2.0   Ta   

   ttop1     D     Ds     Ref  is_sign_given Note Repetitions  
21       -0.87 -0.783  GEH-20              1                   
22        0.59  0.531  GEH-20              1                   
23        0.04  0.032  KHA-16              1                   
24        0.23  0.230  SOU-16              1                   
25        0.12  0.120  SOU-16              1                   
26        0.05  0.050  SOU-16              1                   
27        0.01  0.010  SOU-16              1                   
28        0.68  0.408  JAI-17              1                   
29   5.0  0.03  0.018  JAI-17              1

### Legend 

- **fm**
    : Ferromagnetic layer 
- **tfm**
    : Thickness of ferromagnetic layer (nm)
- **composition**
    : Composition of ferromagnetic layer (206020) for CoFeB means $Co_{20}Fe_{60}B_{20}$
- **bottom1, bottom, top, top1**, **tbottom1, tbottom, ttop, ttop1**, 
    : layers above and over the ferromagnetic layer with their thickness (in nm):   
    line 29 reads $W(5)/Co_{40}Fe_{40}B_{20}(0.6)/MgO(2)/Ta(5)$
- **D, Ds**
    : DMI values, with $D_s = D * t$, where t is tfm
- **Ref**
    : label of the Reference (see below)
- **is_sign_given**
    : 1, True; 0, False
- **Note**
    : Any comment, not used in the code
- **Repetitions**
    : Number of repetions of the stack

# The references

All the references are saved in another Excel file ("dmi_paper.xlsx")

In [5]:
file_refs = 'dmi_paper.xlsx'
refs = pd.read_excel(file_refs, engine='openpyxl',index_col=0)
refs.fillna("").tail()

Author  \
Identifier                                                      
ZHA-15a     Zhang, Vanessa Li; Di, Kai; Lim, Hock Siah; Ng...   
ZHA-18      Zhang, W.; Zhong, H.; Zang, R.; Zhang, Y.; Yu,...   
ZHE-97      Zheludev, A.; Maslov, S.; Shirane, G.; Sasago,...   
ZHE-99      Zheludev, A.; Maslov, S.; Shirane, G.; Tsukada...   
ZIE-18      Ziemys, Grazvydas; Ahrens, Valentin; Mendisch,...   

                                                        Title  \
Identifier                                                      
ZHA-15a     In-plane angular dependence of the spin-wave n...   
ZHA-18      Electrical field enhanced interfacial Dzyalosh...   
ZHE-97      Field-Induced Commensurate-Incommensurate Phas...   
ZHE-99      Magnetic anisotropy and low-energy spin waves ...   
ZIE-18      Speeding up nanomagnetic logic by DMI enhanced...   

                      Journal Volume Number Month         Pages  Year Address  \
Identifier                                                                      
ZHA-15a     Appl. Phys. Lett.    107      2   jul         22402  2015           
ZHA-18       Appl. Phys. Lett    113                     122406  2018           
ZHE-97       Phys. Rev. Lett.     78     25   jun    4857--4860  1997           
ZHE-99           Phys. Rev. B     59          May  11432--11444  1999           
ZIE-18               AIP Adv.      8      5   May         56310  2018           

           Note  
Identifier       
ZHA-15a          
ZHA-18           
ZHE-97           
ZHE-99           
ZIE-18

# How to plot (1): Setting up the variables and initialize the class

To plot the data, we first need to select the methods of measurement we want to show. We must use the shortnames of the table "available_methods" above.

We can also choose to plot the D values or $D_s = D * t$, with $t$ the FM thickness

In [6]:
print(available_methods.index)

Index(['DW_current', 'DW_field_creep', 'DW_field_flow', 'Domain_pattern',
       'Stripe_annihilation', 'Nucleation', 'Stray_field', 'BLS', 'TR_MOKE',
       'PSWS', 'MASW', 'Spin_orbit_torque'],
      dtype='object', name='Method Shortname')


In [7]:
# Methods to be used in the pictures
# If there no data, the method is deleted
methods_to_use = ['DW_field_creep', 'DW_field_flow', 'DW_current', 
           'Domain_pattern', 'Stripe_annihilation', 'Nucleation',
          'BLS', 'PSWS', 'Spin_orbit_torque', 'TR_MOKE', 'MASW']

sel_D = "Ds" # Selection of the DMI value: "D" or "Ds"

# Call the class
dmi = admi.DMI(methods_to_use, filename_DMI, file_refs)

# How to plot (2): Define the composition of the thin film multilayer (the "stack")

A stack is defined as a tuple of 3 lists, as follow:   
stack = ([list of bottom layers], [FM layer], [list of top layers])

NOTE: *One of the two lists of bottom/top layers must be a single element*   
stack_CoFeB = (['IrMn', 'Hf', 'Pt', 'Ta', 'TaN', 'W'], ['CoFeB'], ['MgO'])   
stack_Co = (['Pt'], ['Co'], ['Ir', 'Ta', 'W', 'Pt', 'Au', 'Al', 'Gd', 'GdO', 'AlOx', 'MgO'])

we are now ready to make the plots

# Plot 1: CoFeB with MgO as top layer

Hoove the mouse over the plot to get all the information

In [24]:
stack_CoFeB = (['IrMn', 'Hf', 'Pt', 'Ta', 'TaN', 'W'], ['CoFeB'], ['MgO'])

fig_CoFeB = dmi.plotStack(stack_CoFeB, sel_D, figsizeX=(900,500),
              visualization_library='bokeh', is_XY_to_plot=False, is_anular_to_plot=True)

6 sectors to plot: ['W', 'TaN', 'Ta', 'Pt', 'Hf', 'IrMn']


In [25]:
# Check the references
dmi.print_stack_tables(stack_CoFeB, useJupyter=True)

<IPython.core.display.Latex object>

fm  tfm composition bottom1 tbottom1 bottom  tbottom  top  ttop top1  \
73  CoFeB  2.0      206020                    IrMn      1.0  MgO   2.0   Ta   
74  CoFeB  2.0      206020                    IrMn      2.5  MgO   2.0   Ta   
75  CoFeB  2.0      206020                    IrMn      4.0  MgO   2.0   Ta   
76  CoFeB  2.0      206020                    IrMn      5.0  MgO   2.0   Ta   
77  CoFeB  2.0      206020                    IrMn      6.0  MgO   2.0   Ta   
78  CoFeB  2.0      206020                    IrMn      7.5  MgO   2.0   Ta   

    ttop1      D     Ds    Ref  is_sign_given Note Repetitions Method  
73    2.0 -0.018 -0.036  MA-17              1                     BLS  
74    2.0 -0.061 -0.122  MA-17              1                     BLS  
75    2.0 -0.070 -0.140  MA-17              1                     BLS  
76    2.0 -0.086 -0.172  MA-17              1                     BLS  
77    2.0 -0.100 -0.200  MA-17              1                     BLS  
78    2.0 -0.130 -0.260  MA-17              1                     BLS

<IPython.core.display.Latex object>

fm  tfm composition bottom1 tbottom1 bottom  tbottom  top  ttop top1  \
27  CoFeB  1.0      206020                      Hf      1.0  MgO   2.0        
12  CoFeB  1.0      206020                      Hf      2.6  MgO   2.0        
13  CoFeB  1.0      206020                      Hf      3.5  MgO   2.0        
14  CoFeB  1.0      206020                      Hf      4.7  MgO   2.0        
15  CoFeB  1.0      206020                      Hf      5.9  MgO   2.0        
69  CoFeB  1.0      206020                      Hf      1.0  MgO   2.0   Ta   

   ttop1     D    Ds     Ref  is_sign_given Note Repetitions          Method  
27        0.01  0.01  SOU-16              1                   DW_field_creep  
12       -0.37 -0.37  TOR-14              1                       DW_current  
13       -0.26 -0.26  TOR-14              1                       DW_current  
14       -0.10 -0.10  TOR-14              1                       DW_current  
15       -0.04 -0.04  TOR-14              1                       DW_current  
69   1.0  0.15  0.15  SOU-16              1                              BLS

<IPython.core.display.Latex object>

fm   tfm composition bottom1 tbottom1 bottom  tbottom  top  ttop  \
24   CoFeB  0.80      404020      Ta      3.0     Pt      3.0  MgO   1.5   
25   CoFeB  0.80      602020      Pt      3.7     Pt      2.7  MgO   1.5   
27   CoFeB  0.86      206020      Pt      1.0     Pt      2.7  MgO   1.5   
31   CoFeB  0.80      602020      Ta      3.0     Pt      2.5  MgO   1.5   
4    CoFeB  0.80      404020                      Pt      3.0  MgO   1.5   
5    CoFeB  0.70                                  Pt      3.2  MgO   1.4   
64   CoFeB  0.80      404020     MgO      2.0     Pt      2.0  MgO   2.0   
90   CoFeB  1.00      206020                      Pt      5.0  MgO   2.0   
97   CoFeB  1.00                  Ta      5.0     Pt      0.7  MgO   1.0   
98   CoFeB  1.00                  Ta      5.0     Pt      1.0  MgO   1.0   
99   CoFeB  1.00                  Ta      5.0     Pt      1.7  MgO   1.0   
100  CoFeB  1.00                  Ta      5.0     Pt      2.0  MgO   1.0   
101  CoFeB  1.00                  Ta      5.0     Pt      2.5  MgO   1.0   
102  CoFeB  1.00                  Ta      5.0     Pt      3.0  MgO   1.0   
103  CoFeB  1.00                  Ta      5.0     Pt      4.0  MgO   1.0   
21   CoFeB  1.00                                  Pt      4.0  MgO   2.0   

     top1 ttop1      D      Ds     Ref  is_sign_given Note  \
24     Ta   2.0  1.670  1.3300  WOO-17              0        
25               1.500  1.2000  BUT-17              0        
27     Pt   2.7  1.760  1.5136  LEM-18              0        
31     Ta   2.0  1.600  1.2800  AGR-19              0        
4                1.350  1.0800  WOO-17              0        
5                1.350  0.9450  LIT-17              0        
64   SiO2   3.0 -1.000 -0.8000   DI-15              1        
90     Ta   2.0 -0.970 -0.9700   MA-18              1        
97     Ta   1.0  0.050  0.0500  CHE-18              1        
98     Ta   1.0 -0.087 -0.0870  CHE-18              1        
99     Ta   1.0 -0.130 -0.1300  CHE-18              1        
100    Ta   1.0 -0.320 -0.3200  CHE-18              1        
101    Ta   1.0 -0.550 -0.5500  CHE-18              1        
102    Ta   1.0 -0.767 -0.7670  CHE-18              1        
103    Ta   1.0 -0.830 -0.8300  CHE-18              1        
21               1.800  1.8000  PAI-16              0        

                                          Repetitions               Method  
24                                                 20       Domain_pattern  
25                                                 15       Domain_pattern  
27                                                 15       Domain_pattern  
31                                                 13       Domain_pattern  
4    Ta(3)/[Pt(3)/Co40Fe40B20(0.8)/MgO(1.5)]x20/Ta(2)  Stripe_annihilation  
5              Ta(3)/[Pt(3.2)/CoFeB(0.8)/MgO(1.4)]x15  Stripe_annihilation  
64                                                                     BLS  
90                                                                     BLS  
97                                                                     BLS  
98                                                                     BLS  
99                                                                     BLS  
100                                                                    BLS  
101                                                                    BLS  
102                                                                    BLS  
103                                                                    BLS  
21                                                       Spin_orbit_torque

<IPython.core.display.Latex object>

fm  tfm composition bottom1 tbottom1 bottom  tbottom  top  ttop top1  \
23   CoFeB  0.8      206020                      Ta      5.0  MgO   2.0        
30   CoFeB  0.8      206020                      Ta      5.0  MgO   2.0        
31   CoFeB  1.0      206020                      Ta      5.0  MgO   2.0        
16   CoFeB  1.0      206020                      Ta      0.5  MgO   2.0        
17   CoFeB  1.0      206020                      Ta      0.7  MgO   2.0        
18   CoFeB  1.0      206020                      Ta      1.3  MgO   2.0        
29   CoFeB  1.0      206020                      Ta      5.0  MgO   2.0        
30   CoFeB  0.8      206020                      Ta      5.0  MgO   2.0        
87   CoFeB  1.0      206020                      Ta      5.0  MgO   2.0   Ta   
96   CoFeB  1.0                                  Ta      5.0  MgO   1.0   Ta   
109  CoFeB  1.5      206020                      Ta      1.0  MgO   2.4   Pt   
22   CoFeB  1.0                                  Ta      6.0  MgO   2.0        
24   CoFeB  1.3                                  Ta      3.0  MgO   1.0        

    ttop1      D     Ds      Ref  is_sign_given Note Repetitions  \
23         0.040  0.032   KHA-16              1                    
30         0.030  0.024   KAR-18              1                    
31         0.020  0.020  DIE-19a              1                    
16        -0.070 -0.070   TOR-14              1                    
17        -0.065 -0.065   TOR-14              1                    
18        -0.078 -0.078   TOR-14              1                    
29         0.060  0.060   LOC-15              1                    
30        -0.030 -0.024   KAR-18              1                    
87    2.0  0.040  0.040    MA-18              1                    
96    1.0  0.130  0.130   CHE-18              1                    
109   4.0  0.080  0.000  BEN-20b              1                    
22         0.600  0.600   PAI-16              0                    
24         0.220  0.286  CHE-18a              0                    

                Method  
23      DW_field_creep  
30      DW_field_creep  
31      DW_field_creep  
16          DW_current  
17          DW_current  
18          DW_current  
29          DW_current  
30          DW_current  
87                 BLS  
96                 BLS  
109                BLS  
22   Spin_orbit_torque  
24   Spin_orbit_torque

<IPython.core.display.Latex object>

fm  tfm composition bottom1 tbottom1 bottom  tbottom  top  ttop top1  \
26  CoFeB  1.0      206020                     TaN      1.0  MgO   2.0        
19  CoFeB  1.0      206020                     TaN      0.4  MgO   2.0        
20  CoFeB  1.0      206020                     TaN      1.2  MgO   2.0        
21  CoFeB  1.0      206020                     TaN      2.1  MgO   2.0        
22  CoFeB  1.0      206020                     TaN      3.6  MgO   2.0        
23  CoFeB  1.0      206020                     TaN      5.1  MgO   2.0        
24  CoFeB  1.0      206020                     TaN      6.6  MgO   2.0        
68  CoFeB  1.0      206020                     TaN      1.0  MgO   1.0   Ta   

   ttop1     D    Ds     Ref  is_sign_given Note Repetitions          Method  
26        0.05  0.05  SOU-16              1                   DW_field_creep  
19        0.04  0.04  TOR-14              1                       DW_current  
20        0.17  0.17  TOR-14              1                       DW_current  
21        0.20  0.20  TOR-14              1                       DW_current  
22        0.20  0.20  TOR-14              1                       DW_current  
23        0.19  0.19  TOR-14              1                       DW_current  
24        0.15  0.15  TOR-14              1                       DW_current  
68   1.0  0.31  0.31  SOU-16              1                              BLS

<IPython.core.display.Latex object>

fm  tfm composition bottom1 tbottom1 bottom  tbottom  top  ttop top1  \
24   CoFeB  1.0      206020                       W      2.0  MgO   2.0        
25   CoFeB  1.0      206020                       W      3.0  MgO   2.0        
28   CoFeB  0.6      206020                       W      5.0  MgO   2.0        
29   CoFeB  0.6      404020                       W      5.0  MgO   2.0   Ta   
25   CoFeB  1.0      206020                       W      2.1  MgO   2.0        
26   CoFeB  1.0      206020                       W      2.6  MgO   2.0        
27   CoFeB  1.0      206020                       W      3.1  MgO   2.0        
28   CoFeB  1.0      206020                       W      3.6  MgO   2.0        
3    CoFeB  0.6      206020                       W      5.0  MgO   2.0        
66   CoFeB  1.0      206020                       W      2.0  MgO   1.0   Ta   
67   CoFeB  1.0      206020                       W      3.0  MgO   1.0   Ta   
88   CoFeB  1.0      206020                       W      5.0  MgO   2.0   Ta   
104  CoFeB  0.9      404020                       W     13.0  MgO   1.0   Ta   

    ttop1     D     Ds      Ref  is_sign_given Note Repetitions  \
24         0.23  0.230   SOU-16              1                    
25         0.12  0.120   SOU-16              1                    
28         0.68  0.408   JAI-17              1                    
29    5.0  0.03  0.018   JAI-17              1                    
25         0.51  0.510   TOR-14              1                    
26         0.71  0.710   TOR-14              1                    
27         0.71  0.710   TOR-14              1                    
28         0.47  0.470   TOR-14              1                    
3          0.73  0.438   JAI-17              0                    
66    1.0  0.25  0.250   SOU-16              1                    
67    1.0  0.27  0.270   SOU-16              1                    
88    2.0  0.07  0.070    MA-18              1                    
104   2.0        0.630  KIM-18b              0                    

                  Method  
24        DW_field_creep  
25        DW_field_creep  
28        DW_field_creep  
29        DW_field_creep  
25            DW_current  
26            DW_current  
27            DW_current  
28            DW_current  
3    Stripe_annihilation  
66                   BLS  
67                   BLS  
88                   BLS  
104                  BLS

# Plot 2: Co with Pt as bottom layer

In [10]:
stack_Co = (['Pt'], ['Co'], ['Ir', 'Ta', 'W', 'Pt', 'Au', 'Al', 'Gd', 'GdO', 'AlOx', 'MgO'])
# Play with marker_type = shapes|filling and the alpha

fig_Co = dmi.plotStack(stack_Co, sel_D, figsizeX=(900,500),
              visualization_library='bokeh', is_XY_to_plot=False, is_anular_to_plot=True)

10 sectors to plot: ['MgO', 'AlOx', 'GdO', 'Gd', 'Al', 'Au', 'Pt', 'W', 'Ta', 'Ir']


In [11]:
# Check the references
dmi.print_stack_tables(stack_Co, useJupyter=True)

<IPython.core.display.Latex object>

fm   tfm composition bottom1 tbottom1 bottom  tbottom top  ttop top1  \
2   Co  0.70                                  Pt      5.0  Ir  0.23   Pt   
3   Co  0.70                                  Pt      5.0  Ir  0.69   Pt   
6   Co  0.56                                  Pt      4.0  Ir  5.00        
7   Co  1.05                                  Pt      4.0  Ir  5.00        
17  Co  0.80                  Ta      4.0     Pt      4.0  Ir  4.00        
6   Co  0.60                                  Pt      2.0  Ir  2.00        
2   Co  1.20                                  Pt      4.0  Ir  4.00        
16  Co  2.50                                  Pt      5.4  Ir  2.50   Pt   
29  Co  1.00                  Ta      3.0     Pt      3.0  Ir  3.00        
46  Co  0.80                  Ta      2.0     Pt      2.2  Ir  2.00   Ta   
50  Co  2.50                  Ta      3.0     Pt      3.0  Ir  3.00        
51  Co  2.50                  Ta      3.0     Pt      3.0  Ir  3.00        
0   Co  4.00                  Ta      4.0     Pt      4.0  Ir  3.00   Pt   
7   Co  1.20                                  Pt      4.0  Ir  1.00        
8   Co  1.20                                  Pt      4.0  Ir  2.00        
9   Co  1.00                                  Pt      4.0  Ir  1.00        
15  Co  0.90                  Pt      2.0     Pt      0.6  Ir  0.60  MgO   
17  Co  0.90                  Pt      2.0     Pt      0.6  Ir  0.60  MgO   

   ttop1     D      Ds      Ref  is_sign_given       Note Repetitions  \
2    3.0  0.08  0.0600   HRA-14              1                          
3    3.0 -1.23 -0.8600   HRA-14              1                          
6        -0.31 -0.1736   SHE-18              1                          
7        -0.12 -0.1260   SHE-18              1                          
17       -0.29 -0.2300   HAR-19              1  Co: 0.8-1               
6         -0.5 -0.3000   AJE-17              1                          
2        -1.69 -2.0300   HAN-16              1       Edge               
16   2.6  -0.6 -1.5100   ROW-17              1                          
29        -0.8 -0.8000  BEL-19a              1                          
46   4.0 -0.88 -0.7000   SHA-19              1                          
50        -0.3 -0.7600   BEN-20              1                          
51        -0.2 -0.5300   BEN-20              1                          
0    3.0       -1.0000   LUC-20              1   Co: 4-26               
7         1.09  1.3080   KHA-18              0                          
8         1.42  1.7040   KHA-18              0                          
9         2.06  2.0600   KHA-18              0                          
15   2.0   0.4  0.3600   ISH-19              0                    3.0   
17   2.0   0.7  0.6300   ISH-19              0                          

               Method  
2      DW_field_creep  
3      DW_field_creep  
6      DW_field_creep  
7      DW_field_creep  
17     DW_field_creep  
6       DW_field_flow  
2          Nucleation  
16                BLS  
29                BLS  
46                BLS  
50                BLS  
51                BLS  
0                PSWS  
7   Spin_orbit_torque  
8   Spin_orbit_torque  
9   Spin_orbit_torque  
15  Spin_orbit_torque  
17  Spin_orbit_torque

<IPython.core.display.Latex object>

fm  tfm composition bottom1 tbottom1 bottom  tbottom top  ttop top1 ttop1  \
14  Co  0.8                                  Pt      2.2  Ta   4.0              
3   Co  0.9                  Ta      3.0     Pt      3.0  Ta   4.0              
1   Co  0.9                  Ta      3.0     Pt      3.0  Ta   4.0              
12  Co  2.0                                  Pt      4.0  Ta   4.0              
45  Co  0.8                  Ta      2.0     Pt      2.2  Ta   4.0              
47  Co  1.1                                  Pt      3.0  Ta   4.0              
3   Co  0.5                                  Pt      4.0  Ta   2.0              
4   Co  0.5                                  Pt      4.0  Ta   4.0              
5   Co  0.5                                  Pt      4.0  Ta   6.0              
6   Co  0.5                                  Pt      4.0  Ta   8.0              

       D     Ds     Ref  is_sign_given Note Repetitions               Method  
14 -1.12 -0.900  SHA-19              1                        DW_field_creep  
3   1.50  1.350  WOO-16              0               15       Domain_pattern  
1   1.10  1.000  WOO-16              0               15  Stripe_annihilation  
12 -0.92 -1.840  CHO-17              1                                   BLS  
45 -1.80 -1.640  SHA-19              1                                   BLS  
47 -0.78 -0.860  SAH-19              1               12                  BLS  
3   1.01  0.505  YUN-18              0                     Spin_orbit_torque  
4   0.75  0.375  YUN-18              0                     Spin_orbit_torque  
5   0.70  0.350  YUN-18              0                     Spin_orbit_torque  
6   1.40  0.700  YUN-18              0                     Spin_orbit_torque

<IPython.core.display.Latex object>

fm  tfm composition bottom1 tbottom1 bottom  tbottom top  ttop top1 ttop1  \
10  Co  1.8                                  Pt      5.0   W   1.0   Pt   1.0   
10  Co  0.9                                  Pt      2.5   W   2.5              
33  Co  1.0                                  Pt      1.5   W   1.5              
34  Co  1.0                                  Pt      1.5   W   1.5              
35  Co  1.0                                  Pt      1.5   W   1.5              

       D     Ds      Ref  is_sign_given Note             Repetitions  \
10  0.19  0.342   LIN-18              0                                
10 -1.35 -1.220   KIM-18              1                                
33 -0.55 -0.550  BEN-19a              1                                
34 -0.85 -0.850  BEN-19a              1       [Pt(1.5)Co(1)W(1.5)]x3   
35 -0.90 -0.900  BEN-19a              1       [Pt(1.5)Co(1)W(1.5)]x7   

            Method  
10  DW_field_creep  
10      DW_current  
33             BLS  
34             BLS  
35             BLS

<IPython.core.display.Latex object>

fm   tfm composition bottom1 tbottom1 bottom  tbottom top  ttop top1  \
0   Co  0.30                                  Pt      2.5  Pt   1.5        
1   Co  0.70                                  Pt      5.0  Pt   3.0        
4   Co  0.40                                  Pt      2.0  Pt   2.0        
11  Co  0.60                                  Pt      3.0  Pt   3.0        
15  Co  0.60                  Ta      4.0     Pt      4.0  Pt   4.0        
20  Co  0.90                                  Pt      5.0  Pt   5.0        
4   Co  1.00                                  Pt      4.0  Pt   4.0        
8   Co  0.60                                  Pt      2.0  Pt   2.0        
0   Co  0.36                                  Pt      4.0  Pt   1.0        
1   Co  0.36                                  Pt      4.0  Pt   2.0        
2   Co  0.50                                  Pt      4.0  Pt   2.0        
3   Co  0.50                                  Pt      2.0  Pt   4.0        
11  Co  0.50                                  Pt      2.5  Pt   1.5        
16  Co  0.80                                  Pt      2.5  Pt   1.5        
32  Co  1.00                  Ta      3.0     Pt      3.0  Pt   3.0        
48  Co  1.20                  Ta      2.5     Pt      2.2  Pt   2.0        
49  Co  1.20                  Ta      2.5     Pt      0.4  Pt   2.0        
1   Co  4.00                  Ta      4.0     Pt      4.0  Pt   3.0        

   ttop1     D     Ds      Ref  is_sign_given Note  \
0         0.11  0.033    JE-13              1        
1         0.83  0.581   HRA-14              1        
4         0.33  0.132   KIM-15              1        
11       -0.07 -0.042   SHA-18              1        
15        -1.2 -0.710   HAR-19              1        
20         0.0  0.000   GEH-20              0        
4          0.0  0.000   PHA-16              1        
8          0.0  0.000   AJE-17              0        
0        -0.24 -0.090   FRA-14              1        
1         -0.1 -0.040   FRA-14              1        
2        -0.09 -0.040   FRA-14              1        
3        -0.03 -0.020   FRA-14              1        
11         0.0  0.000   KIM-18              1        
16         0.0  0.000   AGR-19              0        
32        0.08  0.000  BEL-19a              1        
48        0.42  0.500   GUS-20              1        
49        0.27  0.320   GUS-20              1        
1              -0.100   LUC-20              1        

                         Repetitions          Method  
0                                     DW_field_creep  
1                                     DW_field_creep  
4                                     DW_field_creep  
11                                    DW_field_creep  
15                                    DW_field_creep  
20                                    DW_field_creep  
4                                      DW_field_flow  
8                                      DW_field_flow  
0                                         DW_current  
1                                         DW_current  
2                                         DW_current  
3                                         DW_current  
11                                        DW_current  
16  [Pt(2.5-7.5)/Co(0.8)/Pt(1.5)]x13  Domain_pattern  
32                                               BLS  
48                                               BLS  
49                                               BLS  
1                                               PSWS

<IPython.core.display.Latex object>

fm  tfm composition bottom1 tbottom1 bottom  tbottom top  ttop top1 ttop1  \
13  Co  0.6                                  Pt      3.0  Au   3.0              
21  Co  0.9                                  Pt      5.0  Au   5.0              
15  Co  2.5                                  Pt      5.4  Au   2.5   Pt   2.6   

       D     Ds     Ref  is_sign_given Note Repetitions          Method  
13 -1.00 -0.600  SHA-18              1                   DW_field_creep  
21 -0.87 -0.783  GEH-20              1                   DW_field_creep  
15 -0.60 -1.510  ROW-17              1                              BLS

<IPython.core.display.Latex object>

fm  tfm composition bottom1 tbottom1 bottom  tbottom top  ttop top1 ttop1  \
5  Co  0.6                                  Pt      2.0  Al   2.0              
8  Co  0.9                                  Pt      2.5  Al   2.5              

      D     Ds     Ref  is_sign_given Note Repetitions         Method  
5 -1.52 -0.912  AJE-17              1                   DW_field_flow  
8 -0.87 -0.780  KIM-18              1                      DW_current

<IPython.core.display.Latex object>

fm  tfm composition bottom1 tbottom1 bottom  tbottom top  ttop top1 ttop1  \
16  Co  0.8                  Ta      4.0     Pt      4.0  Gd   3.0   Pt   2.0   
18  Co  1.0                  Ta      4.0     Pt      4.0  Gd   3.0   Pt   4.0   
19  Co  1.0                  Ta      4.0     Pt      4.0  Gd   3.0   Ta   4.0   
0   Co  1.0                                  Pt      5.0  Gd   5.0              
3   Co  1.0                                  Pt      4.0  Gd   3.0              
13  Co  1.0                                  Pt      4.0  Gd   4.0              
4   Co  1.0                                  Pt      5.0  Gd   2.0              

       D    Ds      Ref  is_sign_given         Note Repetitions  \
16 -0.42 -0.34   HAR-19              1  Co: 0.8-1.2               
18  0.37  0.37  CAO-20a              0                            
19  0.24  0.24  CAO-20a              0                            
0  -1.48 -1.48   VAN-15              1                            
3  -1.52 -1.52   PHA-16              1                            
13 -1.45 -1.45   KRI-19              1                            
4  -1.00 -1.00   VAN-15              1                            

            Method  
16  DW_field_creep  
18  DW_field_creep  
19  DW_field_creep  
0    DW_field_flow  
3    DW_field_flow  
13   DW_field_flow  
4       DW_current

<IPython.core.display.Latex object>

fm  tfm composition bottom1 tbottom1 bottom  tbottom  top  ttop top1  \
2   Co  1.0                                  Pt      4.0  GdO   4.0        
12  Co  1.0                                  Pt      4.0  GdO   1.0        
14  Co  1.0                                  Pt      4.0  GdO   4.0        
6   Co  1.0                                  Pt      5.0  GdO   2.0   Al   

   ttop1     D    Ds     Ref  is_sign_given Note Repetitions         Method  
2        -1.48 -1.48  PHA-16              1                   DW_field_flow  
12       -0.60 -0.60  SOU-19              1                   DW_field_flow  
14       -1.50 -1.50  KRI-19              1                   DW_field_flow  
6    2.0 -1.60 -1.60  VAN-15              1                             BLS

<IPython.core.display.Latex object>

fm   tfm composition bottom1 tbottom1 bottom  tbottom   top  ttop top1  \
1   Co  0.80                                  Pt      4.0  AlOx   3.0        
10  Co  0.80                                  Pt     30.0  AlOx   1.0        
11  Co  0.80                                  Pt     30.0  AlOx   3.0        
5   Co  0.93                                  Pt      4.0  AlOx   2.0        
6   Co  1.31                                  Pt      4.0  AlOx   2.0        
7   Co  1.37                                  Pt      4.0  AlOx   2.0        
0   Co  0.80                                  Pt      3.0  AlOx   3.0        
0   Co  0.60                                  Pt      3.0  AlOx   2.0        
0   Co  0.60                  Ta      3.0     Pt      3.0  AlOx   2.0   Pt   
1   Co  0.80                  Ta      3.0     Pt      3.0  AlOx   2.0   Pt   
2   Co  0.90                  Ta      3.0     Pt      3.0  AlOx   2.0   Pt   
3   Co  0.95                  Ta      3.0     Pt      3.0  AlOx   2.0   Pt   
4   Co  1.20                  Ta      3.0     Pt      3.0  AlOx   2.0   Pt   
5   Co  2.00                                  Pt      4.0  AlOx   2.0        
7   Co  1.70                                  Pt      4.0  AlOx   2.0        
8   Co  1.80                  Ta      4.0     Pt      4.0  AlOx   2.0        
13  Co  1.60                  Ta      4.0     Pt      0.8  AlOx   2.0   Pt   
14  Co  1.60                  Ta      4.0     Pt      4.8  AlOx   2.0   Pt   
20  Co  0.80                CrPt      8.0     Pt      0.4  AlOx   2.0        

   ttop1      D     Ds     Ref  is_sign_given  Note Repetitions  \
1        -1.630 -1.304  PHA-16              1                     
10       -1.000 -0.800  SOU-19              1                     
11       -1.790 -1.432  SOU-19              1                     
5        -0.540 -0.500  LOC-17              1                     
6        -0.480 -0.630  LOC-17              1                     
7        -0.470 -0.640  LOC-17              1                     
0         0.330  0.270  BEN-15              0                     
0        -2.200 -1.320  PIZ-14              1  Edge               
0    3.0 -2.710 -1.630  BEL-15              1                     
1    3.0 -2.180 -1.750  BEL-15              1                     
2    3.0 -1.880 -1.690  BEL-15              1                     
3    3.0 -1.760 -1.670  BEL-15              1                     
4    3.0 -1.570 -1.884  BEL-15              1                     
5         1.200  1.800  CHO-15              0                     
7        -1.200 -1.600  KIM-15              1                     
8        -1.300 -2.340  KIM-15              1                     
13   1.0 -0.435 -0.696  KIM-17              1                     
14   1.0 -1.570 -2.512  KIM-17              1                     
20        0.870  0.696  QUA-20              0                     

                 Method  
1         DW_field_flow  
10        DW_field_flow  
11        DW_field_flow  
5            DW_current  
6            DW_current  
7            DW_current  
0   Stripe_annihilation  
0            Nucleation  
0                   BLS  
1                   BLS  
2                   BLS  
3                   BLS  
4                   BLS  
5                   BLS  
7                   BLS  
8                   BLS  
13                  BLS  
14                  BLS  
20    Spin_orbit_torque

<IPython.core.display.Latex object>

fm   tfm composition bottom1 tbottom1 bottom  tbottom  top  ttop top1  \
8   Co  1.00                                  Pt      3.0  MgO  0.65   Pt   
3   Co  0.50                                  Pt      2.0  MgO  2.00        
9   Co  1.06                  Ta      3.0     Pt      3.0  MgO  2.00   Ta   
20  Co  1.00                                  Pt      5.0  MgO  2.00   Ta   
31  Co  1.00                  Ta      3.0     Pt      3.0  MgO  1.00        
43  Co  1.40                  Ta      5.0     Pt      5.0  MgO  2.00   Ta   
55  Co  2.20                  Ta      3.0     Pt      3.0  MgO  1.00        
56  Co  2.10                  Ta      3.0     Pt      3.0  MgO  1.00        
0   Co  1.00                                  Pt      4.0  MgO  2.00        
1   Co  0.65                                  Pt      4.0  MgO  2.00        
2   Co  0.65                                  Pt      4.0  MgO  2.00        

   ttop1     D      Ds      Ref  is_sign_given    Note Repetitions  \
8    5.0  0.77  0.7700   CAO-18              0                       
3         0.48  0.2400  KIM-18c              0  Bubble               
9    2.0 -2.05 -2.1730   BOU-16              1                       
20   2.0 -1.59 -1.5900    MA-18              1                       
31       -0.95 -0.9500  BEL-19a              1                       
43   3.0  1.20  1.6800  KIM-19a              0                       
55       -0.50 -0.9500   BEN-20              1                       
56       -0.60 -1.1700   BEN-20              1                       
0         3.00  3.0000   PAI-16              0                       
1         1.45  0.9425   PAI-16              0                       
2         1.99  1.2935   PAI-16              0                       

               Method  
8      DW_field_creep  
3          Nucleation  
9                 BLS  
20                BLS  
31                BLS  
43                BLS  
55                BLS  
56                BLS  
0   Spin_orbit_torque  
1   Spin_orbit_torque  
2   Spin_orbit_torque

# Plot 3:  CoNi with Pt as bottom layer

In [12]:
stack_CoNi = (['Pt'], ['CoNi'], ['MgO', 'Pt', 'Cu', 'Ta', 'Ir'])
fig_CoNi = dmi.plotStack(stack_CoNi, sel_D, figsizeX=(900,500),
              visualization_library='bokeh', is_XY_to_plot=False, is_anular_to_plot=True)

5 sectors to plot: ['Ir', 'Ta', 'Cu', 'Pt', 'MgO']


In [13]:
# Check the references
dmi.print_stack_tables(stack_CoNi, useJupyter=True)

<IPython.core.display.Latex object>

fm  tfm composition bottom1 tbottom1 bottom  tbottom  top  ttop top1  \
32   CoNi  0.9                                  Pt      4.0  MgO   2.0        
113  CoNi  3.2                                  Pt      4.0  MgO   2.0        

    ttop1     D     Ds    Ref  is_sign_given Note Repetitions          Method  
32        -0.05 -0.045  YU-16              1                   DW_field_creep  
113       -0.44 -1.408  DI-15              1                              BLS

<IPython.core.display.Latex object>

fm  tfm composition bottom1 tbottom1 bottom  tbottom top  ttop top1  \
34  CoNi  0.9                                  Pt      4.0  Pt   2.0        

   ttop1    D    Ds    Ref  is_sign_given Note Repetitions          Method  
34        0.2  0.18  YU-16              1                   DW_field_creep

<IPython.core.display.Latex object>

fm  tfm composition bottom1 tbottom1 bottom  tbottom top  ttop top1  \
33  CoNi  0.9                                  Pt      4.0  Cu   2.0        

   ttop1     D     Ds    Ref  is_sign_given Note Repetitions          Method  
33       -0.12 -0.108  YU-16              1                   DW_field_creep

<IPython.core.display.Latex object>

fm  tfm composition bottom1 tbottom1 bottom  tbottom top  ttop top1  \
35  CoNi  0.9                                  Pt      4.0  Ta   2.0        
36  CoNi  1.8                                  Pt      2.5  Ta   0.5  TaN   
37  CoNi  1.8                                  Pt      2.5  Ta   0.5  TaN   
38  CoNi  1.8                                  Pt      1.2  Ta   0.8  TaN   

   ttop1     D     Ds     Ref  is_sign_given Note Repetitions          Method  
35       -0.39 -0.351   YU-16              1                   DW_field_creep  
36   3.0 -0.21 -0.378  LAU-16              1                   DW_field_creep  
37   6.0 -0.37 -0.666  PEL-17              1                   DW_field_creep  
38   6.0 -0.52 -0.936  LAU-18              1                   DW_field_creep

<IPython.core.display.Latex object>

fm  tfm composition bottom1 tbottom1 bottom  tbottom top  ttop top1  \
40  CoNi  1.8                                  Pt      2.5  Ir   2.5        

   ttop1     D     Ds     Ref  is_sign_given Note Repetitions          Method  
40       -0.31 -0.558  LAU-18              1                   DW_field_creep

# Print n most popular stacks

In [15]:
# Print the most popular stacks
q = dmi.print_most_popular_stacks(nfirst=20)
with pd.option_context('display.multi_sparse', False):
    display(q)

top    AlOx  Ir  Pt   MgO   MgO   MgO MgO  Ta Pt   MgO Gd Cu   MgO Ru   MgO  \
fm       Co  Co  Co CoFeB CoFeB CoFeB  Co  Co Co CoFeB Co Co CoFeB Co CoFeB   
bottom   Pt  Pt  Pt    Pt    Ta     W  Pt  Pt Ir   TaN Pt Pt    Hf Pt  IrMn   
n        19  18  18    16    13    13  11  10  8     8  7  6     6  6     6   

top       Pt  W   Ta IrMn GdO  
fm     CoFeB Co CoNi   Co  Co  
bottom    Ta Pt   Pt   Pt  Pt  
n          6  5    4    4   4

In [16]:
# Print the most popular stacks for Co
dmi.print_most_popular_stacks('Co',nfirst=10)

top    AlOx  Ir  Pt MgO  Ta Pt Gd Cu Ru  W
bottom   Pt  Pt  Pt  Pt  Pt Ir Pt Pt Pt Pt
n        19  18  18  11  10  8  7  6  6  5

In [23]:
# Print the most popular stacks for CoFeB
q = dmi.print_most_popular_stacks('CoFeB',nfirst=20)
with pd.option_context('display.multi_sparse', False):
    display(q)

top    MgO MgO MgO MgO  MgO Pt MgO Ti Ta MgO  Ta TaOx TaO   Pt Ta SiO2 Ru AlO  \
bottom  Pt  Ta   W TaN IrMn Ta  Hf Pt Ta  Ir MgO   Ta  Ta TiO2 Pt    W Pt  Pt   
n       16  13  13   8    6  6   6  4  3   2   2    1   1    1  1    1  1   1   

top     Pt   Pt  
bottom MgO SiO2  
n        1    1

In [22]:
# Print the most popular stacks for CoNi
q = dmi.print_most_popular_stacks('CoNi',nfirst=10)
with pd.option_context('display.multi_sparse', False):
    display(q)

top    Ta MgO Cu Ir Pt Pt Ta
bottom Pt  Pt Pt Pt Ir Pt Ir
n       4   2  1  1  1  1  1